gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/tasks/ModuleArgs.java

```diff
@@ -92,23 +100,40 @@ static ImmutableList<String> collectClasspathArgs(
                 .distinct()
                 .sorted()
                 .flatMap(ModuleArgs::addOpensArg);
-        return Stream.concat(exports, opens).collect(ImmutableList.toImmutableList());
-    }
+        Stream<String> enablePreview = enablePreview(javaVersion, parsedJarManifests);
 
-    private static Optional<JarManifestModuleInfo> parseModuleInfo(@Nullable java.util.jar.Manifest jarManifest) {
-        return Optional.ofNullable(jarManifest)
-                .<JarManifestModuleInfo>map(manifest -> JarManifestModuleInfo.builder()
-                        .exports(readManifestAttribute(manifest, ADD_EXPORTS_ATTRIBUTE))
-                        .opens(readManifestAttribute(manifest, ADD_OPENS_ATTRIBUTE))
-                        .build())
-                .filter(JarManifestModuleInfo::isPresent);
+        return Stream.of(exports, opens, enablePreview)
+                .flatMap(Function.identity())
+                .collect(ImmutableList.toImmutableList());
     }
 
-    private static List<String> readManifestAttribute(java.util.jar.Manifest jarManifest, String attribute) {
-        return Optional.ofNullable(
-                        Strings.emptyToNull(jarManifest.getMainAttributes().getValue(attribute)))
-                .map(ENTRY_SPLITTER::splitToList)
-                .orElseGet(ImmutableList::of);
+    private static Stream<String> enablePreview(
+            JavaVersion javaVersion, Map<File, JarManifestModuleInfo> parsedJarManifests) {
+        Map<String, Collection<String>> enablePreviewFromJar = parsedJarManifests.entrySet().stream()
+                .filter(entry -> entry.getValue().enablePreview().isPresent())
+                .collect(Multimaps.toMultimap(
+                        entry -> entry.getValue().enablePreview().get(),
+                        entry -> entry.getKey().getName(),
+                        () -> MultimapBuilder.hashKeys().arrayListValues().build()))
+                .asMap();
```

Comment:
this is where I'd intuitively reach for streamex and an entrystream, but I think we've found that including too many deps in our gradle plugins causes sad times right?